MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs

```diff
@@ -605,30 +633,39 @@ public string GetInputArgument(EncodingJobInfo state, EncodingOptions encodingOp
                 }
 
                 if (state.IsVideoRequest
-                    && string.Equals(encodingOptions.HardwareAccelerationType, "nvenc", StringComparison.OrdinalIgnoreCase)
+                    && string.Equals(options.HardwareAccelerationType, "nvenc", StringComparison.OrdinalIgnoreCase)
                     && isNvdecDecoder)
                 {
                     // Fix for 'No decoder surfaces left' error. https://trac.ffmpeg.org/ticket/7562
                     arg.Append("-hwaccel_output_format cuda -extra_hw_frames 3 -autorotate 0 ");
                 }
 
                 if (state.IsVideoRequest
-                    && ((string.Equals(encodingOptions.HardwareAccelerationType, "nvenc", StringComparison.OrdinalIgnoreCase)
-                         && (isNvdecDecoder || isCuvidHevcDecoder || isSwDecoder))
-                        || (string.Equals(encodingOptions.HardwareAccelerationType, "amf", StringComparison.OrdinalIgnoreCase)
-                            && (isD3d11vaDecoder || isSwDecoder))))
+                    && ((string.Equals(options.HardwareAccelerationType, "nvenc", StringComparison.OrdinalIgnoreCase)
+                         && (isNvdecDecoder || isCuvidHevcDecoder || isCuvidVp9Decoder || isSwDecoder))))
                 {
-                    if (isTonemappingSupported)
+                    if (!isCudaTonemappingSupported && isOpenclTonemappingSupported)
```

Comment:
Maybe we should introduce a [Flags] Enum to have "many" tonemappings supported and then pick one and have if statements like this use that one. Maybe we can rank the tonemappers.

Reply:
I already have ideas in mind. I would like to rewrite the SW/HW pipelines into different classes. Mixing different 
 pipelines and using the booleans anywhere to check are big hassles for maintenance. But before doing this, some related works need to be done first.

1. I wrote a ffcheck cli tool(not finished yet) like the ffdetect via softworkz from emby to probe the HW caps and do SW fallback without unchecking the codecs in the dashboard.

2. Finish the SW tonemapping and AMD AMF zero copy transcoding(opencl filtering).